MSVC/x86 exception handling: rewrite avoiding VC runtime internals

rainers opened the pull request:
- setup exception handler for exceptions during cleanup to avoid C++ calling std::terminate
- remove vcruntime hacks
- when converting exception to error, do not rethrow if the catch handles errors, too
- add support for the "safeseh" attribute

rainers committed:
setup exception handler for exceptions during cleanup to avoid calling std::terminate
remove vcruntime hacks
when converting to exception to error, do not rethrow if the catch handles errors, too

kinke commented on src/ldc/eh/win32.d:
```diff
@@ -260,71 +298,165 @@ private:
 }
 
-extern(C) void _d_leave_cleanup(void* ptr)
+// @safeseh to be marked as "safe" for the OS securtity check
```
securtity => security

kinke commented on src/ldc/eh/win32.d:
```diff
@@ -260,71 +298,165 @@ private:
+}
+
+extern(Windows)
+void RtlUnwind(void *targetFrame, void *targetIp, ExceptionRecord *pExceptRec, void *valueForEAX);
```
`void *` => `void*` etc. pls, for consistency, also for function below

JohanEngelen commented:
Could you also make an IR testcase for this? ;-)
See  https://github.com/JohanEngelen/ldc/blob/8d078f3bd8939dc6a6c3d5209bf63e2f08d81e24/tests/ir/attributes.d
or
https://github.com/ldc-developers/ldc/pull/1244/files

rainers commented:
> Could you also make an IR testcase for this? ;-)

I'll try though I've yet to see any of the IR tests to pass on my system.
Do you think adding an attribute for safeseh is the right direction? It could also be buried in the target-specific attributes, as it is very specific to Win32.

rainers commented:
> securtity => security
> void \* => void\* etc. pls, for consistency, also for function below

Done.

JohanEngelen commented:
_I'll try though I've yet to see any of the IR tests to pass on my system._
Really? Not to hijack this thread: can you open a new issue for it with some details of error messages?

rainers commented:
> Really? Not to hijack this thread: can you open a new issue for it with some details of error messages?

Sorry for being slow with this, I've just reported it here: https://github.com/ldc-developers/ldc/issues/1265

rainers committed:
fibers: implement switching exception information, fix invalid exception chain

rainers commented:
Finally implemented swapping exception information for fibers and figured why exceptions where never caught in fibers.
With this change, core.thread also passes its unittests in debug and release builds.

rainers commented:
I've added support for catching D exceptions in C++, see https://github.com/ldc-developers/dmd-testsuite/pull/15 for a demonstrating test case.

dnadlinger commented on src/core/thread.d:
```diff
@@ -4854,16 +4858,21 @@ private:
             }
 
-            pstack -= EXCEPTION_REGISTRATION.sizeof;
+            // With safeseh(?), the chain must not extend to the very top
```
Thanks for the explanation and the change, it's a lot clearer to me now. (I thought you were referring to the beginning of the record before instead of to the past-the-end address.)

Shouldn't the change also go into the upstream implementation, though? (We can just cherry-pick the fix from master afterwards.)

rainers replied:
> Shouldn't the change also go into the upstream implementation, though? (We can just cherry-pick the fix from master afterwards.)

I don't think dmd will ever support safeseh. It shouldn't do harm to add it, though.

rainers replied:
https://github.com/D-Programming-Language/druntime/pull/1489

rainers commented:
Updated to allow any exception type to be mapped to C++.
I also applied some of the ideas from https://github.com/rainers/druntime/pull/4

kinke commented on src/ldc/eh/win32.d:
```diff
@@ -233,22 +286,41 @@ nothrow:
+    {
+        auto olength = other._length; other._length = _length; _length = olength;
+        auto op      = other._p;      other._p      = _p;      _p = op;
+        auto ocap    = other._cap;    other._cap    = _cap;    _cap = ocap;
```
Just nit-picking, but how about:

``` D
static void swapField(T)(ref T a, ref T b) { T o = b; b = a; a = o; }
swapField(_length, other._length);
swapField(_p,      other._p);
swapField(_cap,    other._cap);
```

rainers replied:
Done. I think there should be a swap function somewhere in druntime, it is implemented again and again...

kinke commented on src/ldc/eh/win32.d:
```diff
@@ -113,103 +129,140 @@ extern(C) void _d_throw_exception(Object e)
 
+__gshared CatchableType* ctCxxDThrowable;
+__gshared CatchableType* ctCxxDException;
+__gshared CatchableType* ctCxxDError;
```
These 3 aren't used anymore, correct?

rainers replied:
Good spot, removed. I wonder if it would be better to just use the C++ mangling for D aswell to avoid the duplication, but the exceptions will look like C++ in the debugger even more (i.e. showing "::" instead ".").

kinke commented:
I've cherry-picked your upstream commit regarding `core.thread`, so another rebase is needed. While at it, could you please squash the last 4 commits to a single one and then factor out the C++ stuff (2nd mangling, `_d_toString()`) in a separate commit again?

rainers committed:
refactorings inspired by https://github.com/rainers/druntime/pull/4

rainers committed:
add support for catching D exception in C++

rainers commented:
Rebased and squashed some commits.

There's still one open issue that doesn't work with this design: cleanup funclets cannot have exception handling themselves, but that's valid in D, e.g. by using try/catch in scope(exit). The LLVM devs don't plan to implement this any time soon.

That's why I want to try outlining cleanup code in the front-end. This might also help the x64 implementation, as it is unlikely that you can setup exception information on x64 as done here (emplace an exception frame in the calling functions stack space).

kinke commented:
> Rebased and squashed some commits.

:+1: 

> That's why I want to try outlining cleanup code in the front-end. This might also help the x64 implementation, as it is unlikely that you can setup exception information on x64 as done here (emplace an exception frame in the calling functions stack space).

And I thought it'd be finished. ;) All the best for your next adventure then, you rock! :)

rainers commented:
Not sure it'll work out this time. I expect some troubles regarding cleanups that are already inside nested functions.
So maybe it's better to get this in as an (almost) working solution for x86...

kinke commented:
> It is unlikely that you can setup exception information on x64 as done here

Do you think the current method (setting up the `std::terminate` hook) is going to be the (first) x64 solution? That's the reason I haven't merged this yet... ;)

rainers commented:
> Do you think the current method (setting up the std::terminate hook) is going to be the (first) x64 solution? That's the reason I haven't merged this yet... ;)

I hope I can try the frontend-outling this weekend. If that doesn't work out, merging the two solutions might be the best solution ATM for x64 support.

rainers commented:
> I hope I can try the frontend-outling this weekend.

FYI: I actually managed to rewrite cleanup blocks, e.g. `finally { code }` to `finally { _d_cleanup(() { code }); }`, with `_d_cleanup` just calling the delegate, but replacing the function for unwind cleanups with another that catches exceptions.

Unfortunately, it causes some of the expected compile errors when having to access several contexts. In addition, errors show the additional indirection in instantiation traces (_d_cleanup must be a template to inherit function attributes). So it doesn't seem a good solution ATM.

I'm currently torn between this and the previous version using the std::terminate hook. Both are rather hacky. I'll investigate how to make the latter safer to avoid skipping a std::terminate call that's not initiated by an exception during cleanup (it would probably also crash).

rainers commented:
Closing this in favor of #60.
